File: LeetCode/17电话号码的字母组合.go
var ans []string

func letterCombinations(digits string) []string {
	if len(digits) == 0 || len(digits) > 4 {
		return nil
	}
	ans = []string{}
	nums := [10]string{"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"}
	backtracking("", digits, 0, nums)
	return ans
}

func backtracking(temp string, digits string, cnt int, nums [10]string) {
	if len(temp) == len(digits) {
		ans = append(ans, temp)
		return
	}
	str := nums[digits[cnt]-'0']
	for i := 0; i < len(str); i++ {
		temp = temp + string(str[i])
		backtracking(temp, digits, cnt+1, nums)
		temp = temp[:len(temp)-1]
	}
}

File: LeetCode/17电话号码的字母组合.java
class Solution {
    List<String> ans;
    public List<String> letterCombinations(String digits) {
        ans = new ArrayList<>();
        if (digits == null || digits.length() == 0) return ans;
        String[] nums = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        backtracking(digits, nums, 0);
        return ans;
    }

    StringBuilder sb = new StringBuilder();
    public void backtracking(String digits, String[] nums, int cnt) {
        if (cnt == digits.length()) {
            ans.add(sb.toString());
            return ;
        }
        String str = nums[digits.charAt(cnt) - '0'];
        for (int i = 0; i < str.length(); i++) {
            sb.append(str.charAt(i));
            backtracking(digits, nums, cnt + 1);
            sb.deleteCharAt(sb.length() - 1);
        }
    }
}